File: android/src/main/java/com/scandit/capacitor/datacapture/core/utils/CapacitorResult.kt
package com.scandit.capacitor.datacapture.core.utils

import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.scandit.datacapture.frameworks.core.result.FrameworksResult
import org.json.JSONObject

class CapacitorResult(private val call: PluginCall) : FrameworksResult {
    override fun success(result: Any?) {
        if (result == null) {
            call.resolve()
            return
        }

        val resultData = if (result is Map<*, *>) {
            JSONObject(result).toString()
        } else {
            result.toString()
        }

        val capacitorPayload = JSObject()
        capacitorPayload.put("data", resultData)

        call.resolve(capacitorPayload)
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        call.reject(errorMessage, errorMessage)
    }
}

class CapacitorNoopResult : FrameworksResult {
    override fun success(result: Any?) {
        // noop
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        // noop
    }
}

File: ios/Sources/ScanditCapacitorCore/CapacitorResult.swift
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2023- Scandit AG. All rights reserved.
 */

import Capacitor
import ScanditFrameworksCore

public struct CapacitorResult: FrameworksResult {
    private let pluginCall: CAPPluginCall

    public init(_ call: CAPPluginCall) {
        pluginCall = call
    }

    public func success(result: Any?) {
        if let resultDict = result as? [String: Any] {
            do {
                let jsonData = try JSONSerialization.data(withJSONObject: resultDict, options: [])
                if let jsonString = String(data: jsonData, encoding: .utf8) {
                    pluginCall.resolve(["data": jsonString])
                }
            } catch {
                reject(code: "JSON_ERROR", message: "Failed to convert to JSON", details: error)
            }
        } else if let unwrappedResult = result {
            let resultAsString = String(describing: unwrappedResult)
            pluginCall.resolve(["data": resultAsString])
        } else {
            pluginCall.resolve()
        }
    }
    
    public func reject(code: String, message: String?, details: Any?) {
        pluginCall.reject(message!, code)
    }
    
    public func reject(error: Error) {
        let message = String(describing: error)
        let nsError = error as NSError
        pluginCall.reject(message, "\(nsError.code)", error)
    }
}
